include/gridtools/fn/backend2/gpu.hpp

```diff
@@ -0,0 +1,159 @@
+                blocks.y = (tuple_util::get<1>(sizes) + threads.y - 1) / threads.y;
+            }
+            if constexpr (ndims_t::value >= 3) {
+                threads.z = block_size_at_dim<BlockSizes, meta::at_c<keys_t, 2>>::value;
+                blocks.z = (tuple_util::get<2>(sizes) + threads.z - 1) / threads.z;
+            }
+            return {blocks, threads};
+        }
+
+        template <class BlockSizes, class Sizes, class StencilStage, class Composite>
+        void apply_stencil_stage(gpu<BlockSizes>, Sizes const &sizes, StencilStage, Composite composite) {
+            auto ptr_holder = sid::get_origin(composite);
+            auto strides = sid::get_strides(composite);
+
+            auto [blocks, threads] = blocks_and_threads<BlockSizes>(sizes);
+            cuda_util::launch(blocks,
+                threads,
+                0,
+                kernel<BlockSizes, decltype(ptr_holder), decltype(strides), StencilStage>,
+                sizes,
+                ptr_holder,
+                strides,
+                StencilStage());
+        }
+
+        template <class ColumnStage, class Seed>
+        struct column_fun_f {
+            Seed m_seed;
+            std::size_t m_v_size;
+
+            template <class Ptr, class Strides>
+            GT_FUNCTION_DEVICE void operator()(Ptr ptr, Strides const &strides) const {
+                ColumnStage()(wstd::move(m_seed), m_v_size, wstd::move(ptr), strides);
+            }
+        };
+
+        template <class Vertical, class BlockSizes, class Sizes, class ColumnStage, class Composite, class Seed>
+        void apply_column_stage(gpu<BlockSizes>, Sizes const &sizes, ColumnStage, Composite composite, Seed seed) {
+            auto ptr_holder = sid::get_origin(composite);
+            auto strides = sid::get_strides(composite);
+            auto h_sizes = hymap::remove_key<Vertical>(sizes);
+            auto v_size = at_key<Vertical>(sizes);
+
+            auto [blocks, threads] = blocks_and_threads<BlockSizes>(h_sizes);
+            cuda_util::launch(blocks,
+                threads,
+                0,
+                kernel<BlockSizes,
+                    decltype(h_sizes),
+                    decltype(ptr_holder),
+                    decltype(strides),
+                    column_fun_f<ColumnStage, Seed>>,
+                h_sizes,
+                ptr_holder,
+                strides,
+                column_fun_f<ColumnStage, Seed>{std::move(seed), v_size});
+        }
+
+        template <class BlockSizes>
+        auto tmp_allocator(gpu<BlockSizes>) {
+            return sid::device::make_cached_allocator(&cuda_util::cuda_malloc<char[]>);
+        }
+
+        template <class T, class BlockSizes, class Allocator, class Sizes>
+        auto allocate_global_tmp(gpu<BlockSizes>, Allocator &allocator, Sizes const &sizes) {
+            return sid::make_contiguous<T, int_t>(allocator, sizes);
```

Comment:
You don't use `StridesKind` parameter, right? I it is intentional, please add somewhere the comment that the contract is that all allocated temporaries that participated in `composite` creation and have the same `Sizes` type should have also `sizes` parameter also the same.

Alternative may to simplify this API a bit:
```c++
template <class BlockSizes, class Sizes>
auto tmp_allocator(gpu<BlockSizes> be, Sizes sizes) {
  return std::tuple(be, sid::device::make_cached_allocator(&cuda_util::cuda_malloc<char[]>), sizes);
}

template <class T, class BlockSizes, class Allocator, class Sizes>
auto allocate_global_tmp(std::tuple<gpu<BlockSizes>, Allocator, Sizes> &alloc) {
  return sid::make_contiguous<T, int_t>(std::get<1>(alloc), std::get<2>(alloc));
}
```

Reply:
> Sorry, I do not remember what `make_contiguous` does exactly. Probably it interprets the first dimension as the innermost. Please verify that it is the desired behavior. Another thought: What should we do in vertical case? Should we move `vertical` dimension to the `outermost`? Or place it just before the `extra` dims?

Yes, the first dimension is innermost (stride `1_c`). If the first one is a horizontal dimension (like it is in non-fn GT), I would expect that this is the right decision for performance for vertical computations as it enables coalesced loads and stores in all cases. But not sure about stencils in unstructured case, as a contiguous vertical dimension there allows for coalesced loads independently of the neighborhood accesses, while a contiguous horizontal one obviously does not. Probably something we have to investigate sooner or later.

Reply:
> You don't use `StridesKind` parameter, right? I it is intentional, please add somewhere the comment that the contract is that all allocated temporaries that participated in `composite` creation and have the same `Sizes` type should have also `sizes` parameter also the same.
> 
> Alternative may to simplify this API a bit:
> 
> ```c++
> template <class BlockSizes, class Sizes>
> auto tmp_allocator(gpu<BlockSizes> be, Sizes sizes) {
>   return std::tuple(be, sid::device::make_cached_allocator(&cuda_util::cuda_malloc<char[]>), sizes);
> }
> 
> template <class T, class BlockSizes, class Allocator, class Sizes>
> auto allocate_global_tmp(std::tuple<gpu<BlockSizes>, Allocator, Sizes> &alloc) {
>   return sid::make_contiguous<T, int_t>(std::get<1>(alloc), std::get<2>(alloc));
> }
> ```

Was not exactly sure here, so I haven’t added everything. I have some open questions:
- Do we also need some extent/halo parameter that is added to the size and used for alignment purposes? (Or alternatively something that is not added to the size but just controls alignment)
- Do we support lower dimensional temporaries? Probably yes? Then of course `StridesKind` is needed, but if it’s only this we can easily choose a different `StridesKind` depending on the number of dimensions.
- Should different run time allocation sizes be supported? Then the API has to include `StridesKind` because we can not derive it from anything at compile time, right?
- Or is it enough to allow for lower dimensional allocations, introduce compile-time extents relative to `sizes` and derive `StridesKind` from the number of dimensions and size of extents?

Reply:
Well, it looks like `StridesKind` topic should be discussed in a zoom session -- too many open questions.
Here some factors to consider before that discussion:
 - If you don't want to deal with stride kinds correctness use 'sid::unkonown_kind'.
 - We do not control the StrideKInds of the users fields in this API. This means that strides kinds should of the temporaries should be different from any others. 
 - `make_contiguous` has some simple logic for setting default strides kinds (see implementation) -- strides kinds depends on the `Sizes` type. This means that lower dimensional temporaries will have different stride kind comparing to the full dimensional.
 - observation: if a global temporary have extents against domain size it is likely should be block temporary